core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -0,0 +1,210 @@
+case class BrokerToControllerQueueItem(request: AbstractRequest.Builder[_ <: AbstractRequest],
+                     callback: RequestCompletionHandler)
+
+class BrokerToControllerRequestThread(networkClient: KafkaClient,
+                                      metadataUpdater: ManualMetadataUpdater,
+                                      requestQueue: BlockingQueue[BrokerToControllerQueueItem],
+                                      metadataCache: MetadataCache,
+                                      config: KafkaConfig,
+                                      listenerName: ListenerName,
+                                      time: Time,
+                                      threadName: String)
+  extends InterBrokerSendThread(threadName, networkClient, time, isInterruptible = false, exitOnError = false) {
+
+  private var activeController: Option[Node] = None
+
+  override def requestTimeoutMs: Int = config.controllerSocketTimeoutMs
+
+  override def generateRequests(): Iterable[RequestAndCompletionHandler] = {
+    val requestsToSend = new mutable.Queue[RequestAndCompletionHandler]
+    while (requestQueue.peek() != null) {
+      val topRequest = requestQueue.take()
+      val request = RequestAndCompletionHandler(activeController.get, topRequest.request, handleResponse(topRequest))
+      requestsToSend.enqueue(request)
+    }
+    requestsToSend
+  }
+
+  private[server] def handleResponse(request: BrokerToControllerQueueItem)(response: ClientResponse): Unit = {
+    if (response.wasDisconnected()) {
+      activeController = None
+      requestQueue.put(request)
+    } else if (response.responseBody().errorCounts().containsKey(Errors.NOT_CONTROLLER)) {
+      // just close the controller connection and wait for metadata cache update in doWork
+      networkClient.close(activeController.get.idString)
+      activeController = None
+      requestQueue.put(request)
+    } else {
+      request.callback.onComplete(response)
+    }
+  }
+
+  private[server] def backoff(): Unit = pause(100, TimeUnit.MILLISECONDS)
+
+  override def doWork(): Unit = {
+    try {
+      if (activeController.isDefined) {
+        super.doWork()
+      } else {
+        info("Controller isn't cached, looking for local metadata changes")
+        val controllerOpt = metadataCache.getControllerId.flatMap(metadataCache.getAliveBroker)
+        if (controllerOpt.isDefined) {
+          if (activeController.isEmpty || activeController.exists(_.id != controllerOpt.get.id))
+            info(s"Recorded new controller, from now on will use broker ${controllerOpt.get.id}")
+          activeController = Option(controllerOpt.get.node(listenerName))
+          metadataUpdater.setNodes(metadataCache.getAliveBrokers.map(_.node(listenerName)).asJava)
+        } else {
+          // need to backoff to avoid tight loops
+          warn("No controller defined in metadata cache, retrying after backoff")
+          backoff()
+        }
+      }
+    } catch {
+      case e: Exception =>
```

Comment:
Should we be more strict here about exception handling? I don't think we shall continue in every possible exceptions,  if we could brainstorm :)

Reply:
Of course, let's brainstorm! :)

My main goal with this was that ideally we should catch network related exceptions when the request fails due to disconnect events. Do you have specific ideas about what to catch here?

Reply:
Yea, I agree. We could just stay here as long as we are not detecting any other fatal exceptions.

Reply:
Tomorrow I'll write some test cases for this for some specific scenarios.
I think though that usually we should try and handle most exceptions and reconnect to a controller when possible. On the other hand we likely don't want to catch any non-Exception Throwables as those are usually more serious cases (for instance OOM).